File: solutions/typescript/rational-numbers/1/rational-numbers.ts
function gcd(a: number, b: number): number {
  return b ? gcd(b, a % b) : a;
}

function nthRoot(base: number, n: number): number {
  return Math.pow(base, 1 / n);
}

export class Rational {
  numerator: number;
  denominator: number;

  constructor(numerator: number, denominator: number) {
    if (denominator === 0) {
      throw "Denominator value must be != 0";
    }
    this.numerator = numerator;
    this.denominator = denominator;
    this.reduce();
  }

  add(other: Rational): Rational {
    return new Rational(
      this.numerator * other.denominator + this.denominator * other.numerator,
      this.denominator * other.denominator
    );
  }

  sub(other: Rational): Rational {
    return this.add(new Rational(-other.numerator, other.denominator));
  }

  mul(other: Rational): Rational {
    return new Rational(this.numerator * other.numerator, this.denominator * other.denominator);
  }

  div(other: Rational): Rational {
    return new Rational(this.numerator * other.denominator, this.denominator * other.numerator);
  }

  abs(): Rational {
    return new Rational(Math.abs(this.numerator), Math.abs(this.denominator));
  }

  exprational(exp: number): Rational {
    return exp > 0
      ? new Rational(this.numerator ** exp, this.denominator ** exp)
      : new Rational(this.denominator ** -exp, this.numerator ** -exp);
  }

  expreal(base: number): number {
    return nthRoot(base, this.denominator) ** this.numerator;
  }

  reduce(): Rational {
    const reduced = gcd(this.numerator, this.denominator);
    this.numerator /= reduced;
    this.denominator /= reduced;
    if (this.denominator < 0) {
      this.numerator *= -1;
      this.denominator *= -1;
    }
    return this;
  }
}

File: solutions/elixir/rational-numbers/1/lib/rational_numbers.ex
defmodule RationalNumbers do
  @type rational :: {integer, integer}

  @doc """
  Add two rational numbers
  """
  @spec add(a :: rational, b :: rational) :: rational
  def add({a1, b1}, {a2, b2}) do
    {(a1 * b2 + a2 * b1), (b1 * b2)}
    |> reduce
  end

  @doc """
  Subtract two rational numbers
  """
  @spec subtract(a :: rational, b :: rational) :: rational
  def subtract({a1, b1}, {a2, b2}) do
    {(a1 * b2 - a2 * b1), (b1 * b2)}
    |> reduce
  end

  @doc """
  Multiply two rational numbers
  """
  @spec multiply(a :: rational, b :: rational) :: rational
  def multiply({a1, b1}, {a2, b2}) do
    {(a1 * a2), (b1 * b2)}
    |> reduce
  end

  @doc """
  Divide two rational numbers
  """
  @spec divide_by(num :: rational, den :: rational) :: rational
  def divide_by({a1, b1}, {a2, b2}) do
    {(a1 * b2), (a2 * b1)}
    |> reduce
  end

  @doc """
  Absolute value of a rational number
  """
  @spec abs(a :: rational) :: rational
  def abs({a, b}) do
    {Kernel.abs(a), Kernel.abs(b)}
    |> reduce
  end

  @doc """
  Exponentiation of a rational number by an integer
  """
  @spec pow_rational(a :: rational, n :: integer) :: rational
  def pow_rational({a, b}, n) when n >= 0 do
    {:math.pow(a,n) |> round, :math.pow(b,n) |> round}
    |> reduce
  end
  def pow_rational({a, b}, n) when n < 0 do
    m = Kernel.abs(n)
    {:math.pow(b,m) |> round, :math.pow(a,m) |> round}
    |> reduce
  end

  @doc """
  Exponentiation of a real number by a rational number
  """
  @spec pow_real(x :: integer, n :: rational) :: float
  def pow_real(x, {a, b}) do
    x ** (a/b)
  end

  @doc """
  Reduce a rational number to its lowest terms
  """
  @spec reduce(a :: rational) :: rational
  def reduce({0, _b}), do: {0, 1}
  def reduce({a, b}) do
    divisor = gcd(a, b)
    {numerator, denominator} = {div(a, divisor), div(b, divisor)}
    if denominator < 0 do
      {numerator * -1, denominator * -1}
    else
      {numerator, denominator}
    end
  end

  defp gcd(x, 0), do: x
  defp gcd(x, y), do: gcd(y, rem(x,y))
end
